python/semantic_kernel/kernel.py

```diff
@@ -408,51 +297,121 @@ async def invoke_prompt(
     # endregion
     # region Function Invoking/Invoked Events
 
-    def on_function_invoked(
+    async def pre_function_invoke(
+        self,
+        kernel_function_metadata: KernelFunctionMetadata,
+        arguments: KernelArguments,
+        metadata: dict[str, Any] = {},
+    ) -> PreFunctionInvokeContext:
+        context = PreFunctionInvokeContext(
+            kernel_function_metadata=kernel_function_metadata, arguments=arguments, metadata=metadata
+        )
+        for hook in self.hooks:
+            try:
+                if isinstance(hook, HOOK_PROTOCOLS["pre_function_invoke"]):
+                    logger.debug(f"Running Post Function Invoke Hook: {hook.__class__.__name__}")
+                    result = hook.pre_function_invoke(context=context)
+                    if inspect.isawaitable(result):
+                        await result
+            except Exception as exc:
+                logger.error(
+                    f"An error occurred in the on_function_invoking event of hook: {hook.__class__.__name__}. "
+                    f"Error description: {str(exc)}"
+                )
+            if context.is_cancel_requested:
+                raise OperationCancelledException(
+                    f"Execution was cancelled on function invoking event of pipeline step "
+                    f"{metadata.get('pipeline_step', '') if metadata else ''}: "
+                    f"{kernel_function_metadata.fully_qualified_name}."
+                )
+            if context.updated_arguments:
+                logger.info(
+                    f"Arguments updated by function_invoking_handler in pipeline step: "
+                    f"{metadata['pipeline_step']}, new arguments: {context.arguments}"
+                )
+            if context.is_skip_requested:
+                logger.info(
+                    f"Execution was skipped on function invoking event of pipeline step "
+                    f"{metadata.get('pipeline_step', '') if metadata else ''}: "
+                    f"{kernel_function_metadata.fully_qualified_name}."
+                )
+        return context
```

Comment:
I think if there are no hooks registered, this method should return `None` context. In this case, it should be easier to identify in our code when hook was triggered and how to handle it.

Reply:
good one, done